network_simulator/flow.py

```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+
+        # Go through the queue and remove any previously scheduled
+        # FlowTimeoutPacketEvent of the packets in packets_to_send.
+        new_queue = PriorityQueue()
+        while not self.events.empty():
+            curr_tup = self.events.get()
+            curr_event = self.curr_tup[1]
+            if type(curr_event) is FlowTimeoutPacketEvent \
+               and curr_event.packet in self.packets_to_send:
+                continue
+            new_queue.add(curr_tup)
+
+        # Schedule FlowTimeoutPacketEvent for each packet, which will occur
+        # in FLOW_TIMEOUT_SEC from now.
+        for curr_packet in self.packets_to_send:
+            timeout_event = FlowTimeoutPacketEvent(self.flow, curr_packet)
+            main_event_loop.schedule_event_with_delay(timeout_event,
+                                                      FLOW_TIMEOUT_SEC)
+
+
+
+class FlowTimeoutPacketEvent(Event):
+    def __init__(self, flow, packet):
+        """
+        Triggered when packet has timeout at a flow.
+
+        :ivar Flow flow: flow of this event.
+        :ivar packet packet: the particular packet of this timeout.
+        """
+        self.flow = flow
+        self.packet = packet
+
+    def run(self, main_event_loop, statistics):
+        # Does nothing.
+        pass
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedule DeviceToLinkEvent and renew FlowTimeoutPacketEvent for each
+        packet.
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        # If the particular packet is currently in transit, then we don't
+        # need to do anything here.
+        if self.packet.packet_id not in self.flow.packets_in_transit:
+            return
+        flow_send_event = FlowSendPacketsEvent(self.flow, [self.packets])
+        main_event_loop.schedule_event_with_delay(flow_send_event, 0.0)
+
+
+
+class FlowReceivedAckEvent(Event):
+    """
+    Triggered when an ACK is received by a flow.
+    """
+    def __init__(self, flow, packet):
+        """
+        :ivar Flow flow: flow of this event.
+        :ivar AckPacket packet: ack packet received.
+        """
+        self.flow = flow
+        self.packet = packet
+
+    def run(self, main_event_loop, statistics):
+        """
+        Triggered whenever an ACK is received.
+
+        :param MainEventLoop main_event_loop: main loop to retrieve the
+        simulation time of function call.
+        :param Statistics statistics: the Statistics to update.
+        """
+        # Update RTT.
+        sent_time = self.packet.data_packet_start_time_sec
+        curr_rtt = main_event_loop.global_clock_sec - sent_time
+        self.flow.packet_rtts.append((self.packet.packet_id, curr_rtt))
+
+        # Remove packet in transit.
+        self.flow.packets_in_transit.remove(self.packet.packet_id)
+
+        # TODO(sharon): Base on following description, update the code when
```

Comment:
The first thing I would do is flow.handle_packet_success() on the packet that was ACK'd. This will at least set in motion the initial window size update based on successful transmission (exact specifics depend on TCP algo).

If packet loss occurs (cases below), then handle_packet_loss(). This is a less common case so should happen after.